File: backend/src/main/scala/domains/usergroup/http/api/ListUserGroups.scala
package domains.usergroup.http.api

import cats.effect.IO
import domains.auth.http.AuthenticatedApi
import domains.auth.objects.AuthUser
import domains.usergroup.http.codec.UserGroupHttpCodecs.given
import domains.usergroup.objects.response.UserGroupSummary
import domains.usergroup.rules.UserGroupAccessRules
import domains.usergroup.table.user_group.UserGroupTable
import io.circe.Encoder
import org.http4s.{Method, Request, Status}
import shared.http.utils.PageRequestQuerySupport
import shared.http.{ApiPath, PathParams}
import shared.objects.{PageRequest, PageResponse}

import java.sql.Connection

object ListUserGroups extends AuthenticatedApi[PageRequest, PageResponse[UserGroupSummary]]:

  override val method: Method = Method.GET
  override val path: ApiPath = ApiPath("/api/user-groups")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[PageResponse[UserGroupSummary]] = summon[Encoder[PageResponse[UserGroupSummary]]]

  override def decode(request: Request[IO], pathParams: PathParams): IO[PageRequest] =
    val _ = pathParams
    IO.pure(PageRequestQuerySupport.parsePageRequest(request.uri.query.params))

  override def plan(connection: Connection, actor: AuthUser, pageRequest: PageRequest): IO[PageResponse[UserGroupSummary]] =
    val normalizedPageRequest = pageRequest.normalized
    if !UserGroupAccessRules.canList(actor) then
      IO.pure(PageResponse(items = Nil, page = normalizedPageRequest.page, pageSize = normalizedPageRequest.pageSize, totalItems = 0L))
    else
      UserGroupTable.listVisibleTo(connection, actor, normalizedPageRequest.page, normalizedPageRequest.pageSize)

File: frontend/src/apis/usergroup/ListUserGroups.ts
import type { APIMessage } from '@/system/api/api-message'
import type { PageRequest } from '@/objects/shared/PageRequest'
import type { UserGroupListResponse } from '@/objects/usergroup/response/UserGroupListResponse'

/** 查询用户组列表；可选分页参数，输出用户组摘要分页响应。 */
export class ListUserGroups implements APIMessage<UserGroupListResponse> {
  declare readonly responseType?: UserGroupListResponse
  readonly method = 'GET'
  readonly apiPath: string

  constructor(pageRequest?: PageRequest) {
    const params = new URLSearchParams()
    if (pageRequest) {
      params.set('page', String(pageRequest.page))
      params.set('pageSize', String(pageRequest.pageSize))
    }
    this.apiPath = `user-groups${params.size > 0 ? `?${params.toString()}` : ''}`
  }

  body(): undefined {
    return undefined
  }
}
